src/main/diagnostics/diagnostics.ts

```diff
@@ -1,68 +1,121 @@
-import DIAGNOSTICS_TABLE from '@/assets/diagnostics.yaml';
+export enum DiagnosticsCategory {
+  Utilities = 'Utilities',
+  Networking = 'Networking',
+}
+
+/**
+ * DiagnosticsChecker describes an implementation of a single diagnostics check.
+ */
+export interface DiagnosticsChecker {
+  /** Unique identifier for this check. */
+  id: string;
+  documentation: string,
+  description: string,
+  category: DiagnosticsCategory,
+  /**
+   * Preform the check.  Returns true if the check passed (i.e. things are
+   * working correctly and does not need to be fixed).
+   */
+  check(): Promise<boolean>;
+}
 
 type DiagnosticsFix = {
-  fixes: { description: string }
+  description: string
 };
 
-export type DiagnosticsCheck = {
+export type DiagnosticsResult = {
   id: string,
   documentation: string,
   description: string,
-  category: string,
+  category: DiagnosticsCategory,
+  passed: boolean,
   mute: boolean,
   fixes: DiagnosticsFix[],
 };
 
-type DiagnosticsCategories = {
-  title: string,
-  checks: Array<DiagnosticsCheck>,
-};
-
-type DiagnosticsType = {
-  last_update: string
-  categories: Array<DiagnosticsCategories>,
+export type DiagnosticsResultGroup = {
+  last_update: string,
+  checks: Array<DiagnosticsResult>,
 };
 
 export class Diagnostics {
-  diagnostics: DiagnosticsType;
-  checksByCategory: Record<string, Array<DiagnosticsCheck>> = {};
-  checks: Array<DiagnosticsCheck> = [];
-  constructor(diagnosticsTable: DiagnosticsType|undefined = undefined) {
-    this.diagnostics = diagnosticsTable || DIAGNOSTICS_TABLE.diagnostics;
-    for (const category of this.diagnostics.categories) {
-      for (const check of category.checks) {
-        check.mute ??= false;
-        check.fixes ??= [];
-        check.category = category.title;
-        this.checks.push(check);
+  /** Checkers capable of running individual diagnostics. */
+  checkers: Promise<DiagnosticsChecker[]>;
+
+  /** Time stamp of when the last check occurred. */
+  lastUpdate = new Date(0);
+
+  /** Last known check results. */
+  results: Array<DiagnosticsResult> = [];
+
+  /** Mapping of category name to diagnostic ids */
+  categories: Partial<Record<DiagnosticsCategory, string[]>> = {};
+
+  constructor(diagnostics?: DiagnosticsChecker[]) {
```

Comment:
That… makes so much more sense.

Reply:
This turns out to not work as well, since we essentially have to cache the results (when you go to the page, it should display the last results instead of waiting to fetch new results, because running tests can take a while).  Also, it doesn't make sense to return id + category from the checker again (because that provides a chance of getting mismatches).  Finally, the muting mechanism should not be handled in each checker.

I've made the checkers return more things though.